bpf/lib/ratelimit.h

```diff
@@ -0,0 +1,76 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#ifndef __RATELIMIT_H_
+#define __RATELIMIT_H_
+
+#include "common.h"
+#include "bpf/helpers.h"
+
+struct ratelimit_key {
+	__u32 netdev_idx;
+};
+
+struct ratelimit_value {
+	__u64 last_topup;
+	__u64 tokens;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__type(key, struct ratelimit_key);
+	__type(value, struct ratelimit_value);
+	__uint(pinning, LIBBPF_PIN_BY_NAME);
+	__uint(max_entries, 1024);
+} RATELIMIT_MAP __section_maps_btf;
+
+struct ratelimit_settings {
+	/* A bucket will never have more than X amount of tokens, limits burt size */
+	__u64 bucket_size;
+	/* The amount of tokens added to a bucket for every topup */
+	__u64 tokens_per_topup;
+	/* The interval at which the topups happen */
+	__u64 topup_interval_ns;
+};
+
+static inline bool ratelimit_check_and_take(struct ratelimit_key *key,
+					    const struct ratelimit_settings *settings)
+{
+	struct ratelimit_value *value;
+	struct ratelimit_value new_value;
+	__u64 since_last_topup;
+	int ret;
+
+	/* Create a new bucket if we do not yet have one for the key */
+	value = map_lookup_elem(&RATELIMIT_MAP, key);
+	if (!value) {
+		new_value.last_topup = ktime_get_ns();
+		new_value.tokens = settings->tokens_per_topup - 1;
+		ret = map_update_elem(&RATELIMIT_MAP, key, &new_value, BPF_ANY);
```

Comment:
I considered this. It would mean that the ratelimit isn't 100% accurate, letting more traffic through than the limit. To fix that we would need to use atomics which are slow. I thought the performance is more important in this given situation. Perhaps we should add this to the comments in case other wonder.